utils_cv/action_recognition/dataset.py

```diff
@@ -0,0 +1,473 @@
+                    if step == 0:
+                        clip.append(clip[-1].copy())
+                    else:
+                        if step > 1:
+                            video_reader.skip_frames(step - 1)
+                        cur_frame = video_reader.next().asnumpy()
+                        if len(cur_frame.shape) != 3:
+                            # maybe end of the video
+                            break
+                        clip.append(cur_frame)
+            else:
+                for i in range(self.sample_length - 1):
+                    if self.sample_step > 1:
+                        video_reader.skip_frames(self.sample_step - 1)
+                    cur_frame = video_reader.next().asnumpy()
+                    if len(cur_frame.shape) != 3:
+                        # maybe end of the video
+                        break
+                    clip.append(cur_frame)
+        except StopIteration:
+            pass
+
+        # if clip needs more frames, simply duplicate the last frame in the clip.
+        while len(clip) < self.sample_length:
+            clip.append(clip[-1].copy())
+
+        return clip
+
+    def __getitem__(self, idx: int) -> Tuple[torch.tensor, int]:
+        """
+        Return:
+            clips (torch.tensor), label (int)
+        """
+        record = self.video_records[idx]
+        video_reader = decord.VideoReader(
+            "{}.{}".format(
+                os.path.join(self.root, record.path), self.video_ext
+            ),
+            # TODO try to add `ctx=decord.ndarray.gpu(0) or .cuda(0)`
+        )
+        record._num_frames = len(video_reader)
+
+        offsets = self._sample_indices(record)
+        clips = np.array([self._get_frames(video_reader, o) for o in offsets])
+
+        if self.num_segments == 1:
+            # [T, H, W, C] -> [C, T, H, W]
+            return self.transforms(torch.from_numpy(clips[0])), record.label
+        else:
+            # [S, T, H, W, C] -> [S, C, T, H, W]
+            return (
+                torch.stack(
+                    [self.transforms(torch.from_numpy(c)) for c in clips]
+                ),
+                record.label,
+            )
+
+    def _show_batch(
+        self,
+        batch: List[torch.tensor],
+        sample_length: int,
+        mean: Tuple[int, int, int] = DEFAULT_MEAN,
+        std: Tuple[int, int, int] = DEFAULT_STD,
+    ) -> None:
+        """
+        Display a batch of images.
+
+        Args:
+            batch: List of sample (clip) tensors
+            sample_length: Number of frames to show for each sample
+            mean: Normalization mean
+            std: Normalization std-dev
+        """
+        batch_size = len(batch)
+        plt.tight_layout()
+        fig, axs = plt.subplots(
+            batch_size,
+            sample_length,
+            figsize=(4 * sample_length, 3 * batch_size),
+        )
+
+        for i, ax in enumerate(axs):
+            if batch_size == 1:
```

Comment:
Is this line necessary, and if so why?

Reply:
unfortunately this is necessary bc we're iterating through the axs which when rows=1, is the size of the number of columns. But when rows>1,  is the size of the rows.